localstack/services/apigateway/provider_asf.py

```diff
@@ -0,0 +1,191 @@
+        api_id = rest_api["id"]
+
+        # add the canonical execute-api handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/<path:path>",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+
+        # add the localstack-specific _user_request_ handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_/<path:path>",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+
+    def remove_rest_api(self, rest_api_id: str) -> None:
+        """
+        Removes the given rest api.
+        :param rest_api_id: the rest api to remove
+        """
+        rules = self.router_rules.pop(rest_api_id, [])
+        for rule in rules:
+            self.router.remove_rule(rule)
+
+    def _restapis_handler(self, request: Request, path=None) -> Response:
+        invocation_context = to_invocation_context(request)
+
+        result = invoke_rest_api_from_request(invocation_context)
+        if result is not None:
+            if isinstance(result, LambdaResponse):
+                headers = Headers(dict(result.headers))
+                for k, values in result.multi_value_headers.items():
+                    for value in values:
+                        headers.add(k, value)
+            else:
+                headers = dict(result.headers)
+            return Response(
+                response=result.content,
+                status=result.status_code,
+                headers=headers,
+            )
+
+        raise NotFound()
+
+    def _restapis_user_request_handler(self, request: Request, stage=None, path=None):
+        return self._restapis_handler(request, path)
+
+    def _restapis_host_handler(self, request: Request, path=None, server=None) -> Response:
+        return self._restapis_handler(request)
+
+
+class AsfApigatewayProvider(ApigatewayProvider):
+    """Modern ASF provider that uses an ApigatewayRouter to dispatch requests to user routes."""
+
+    router: ApigatewayRouter
+
+    def __init__(self, router: ApigatewayRouter = None):
+        self.router = router or ApigatewayRouter(router=ROUTER)
+
+    def create_rest_api(self, context: RequestContext, request: CreateRestApiRequest) -> RestApi:
+        result: RestApi = super().create_rest_api(context, request)
+        API_REGIONS[result["id"]] = context.region
```

Comment:
Not an ask for change, more like a general comment around the `API_REGIONS` singleton - we have been trying to remove it and we already do the lookup using different approaches. In conclusion, not only we are spreading the use of the singleton here, we were aiming for a lookup on the model object. Again, just a heads-up, we are doing something similar already [here](https://github.com/localstack/localstack/blob/235a8b377870426bcc1f12bb9e2973bcce8f74cf/localstack/services/apigateway/provider.py#L117)

Reply:
that's OK i just retained the old behavior. i'm happy to get rid of it :-) i just didn't understand the implications. maybe we can do a second iteration